File: study/week14/Service/BadThreadExam/app/src/main/java/com/sesong/badthreadexam/MainActivity.java
package com.sesong.badthreadexam;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

public class MainActivity extends AppCompatActivity {
    // 스레드 객체
    private Thread mThread;
    // 카운팅
    private int mCount = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    // 스레드 시작
    public void startThread(View view) {
        if (mThread == null) {
            // 스레드 초기화 및 시작
            mThread = new Thread("My Thread") {
                @Override
                public void run() {
                    for (int i = 0; i < 100; i++) {
                        try {
                            mCount++;
                            // 1초마다 쉬기
                            Thread.sleep(1000);
                        } catch (InterruptedException e) {
                            // 인터럽트되면 오래 걸리는 처리 종료
                            break;
                        }
                        // 1초마다 로그 남기기
                        Log.d("My Thread", "스레드 동작 중" + mCount);
                    }
                }
            };
            mThread.start();
        }
    }

    // 스레드 종료
    public void stopThread(View view) {
        if (mThread != null) {
            mThread.interrupt();
            mThread = null;
            mCount = 0;
        }
    }
}

File: study/week14/Socket/Kotlin Socket/app/src/main/java/com/jinwoo/socket/MainActivity.kt
package com.jinwoo.socket

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import io.socket.client.Socket
import io.socket.emitter.Emitter
import org.json.JSONException

class MainActivity : AppCompatActivity() {

    lateinit var Text : TextView
    lateinit var Receive_Text : TextView
    lateinit var connect_btn: Button
    lateinit var disconnect_btn: Button
    lateinit var socket: Socket
    lateinit var receive_data : String
    lateinit var receive_btn: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        connect_btn = findViewById(R.id.connect_button)
        disconnect_btn = findViewById(R.id.disconnect_button)
        Text = findViewById(R.id.message)
        Receive_Text = findViewById(R.id.receive_Text)
        receive_btn = findViewById(R.id.receive_btn)

        socket = SocketApplication.get()
        socket.connect()

        if(socket != null){
            Text.setText("Socket 생성")
        }



        connect_btn.setOnClickListener { v ->
            Thread(Runnable {
                socket.on("Connecting", onConnect)
            }).start()
        }

        receive_btn.setOnClickListener { v ->
            Thread(Runnable {
                socket.emit("call")
                socket.on("data_call", Receiver)
                Text.setText("Receive 성공")
            }).start()
        }

        disconnect_btn.setOnClickListener { v ->
            Thread(Runnable {
                socket.disconnect()
                socket.off(Socket.EVENT_DISCONNECT)
                Text.setText("Disconnect 성공")
            }).start()
        }
    }

    private val onConnect = Emitter.Listener { args ->

            var data = "Hello Server"

            try {
                socket.emit("Connect", data)
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        Text.setText("Connect 성공")
    }

    private val Receiver = Emitter.Listener { args ->

        try{
            receive_data = args.toString()
        } catch( e : JSONException){
            e.printStackTrace()
        }
        Receive_Text.setText(receive_data)
    }
}
